mm/pmm.c

```diff
@@ -108,7 +141,49 @@ static frames_array_t *new_frames_array(void) {
     if (!boot_flags.virt)
         array = (frames_array_t *) mfn_to_virt_kern(frame->mfn);
     else {
+        /* switch to special refilling mode to avoid deadlock with paging */
+        spin_unlock(&priority_lock);
+
+        /* only paging is allowed to use pmm with only the priority_lock */
         array = vmap_kern_4k(mfn_to_virt_map(frame->mfn), frame->mfn, L1_PROT);
+
+        /* switch back to normal mode */
+        spin_lock(&priority_lock);
+
+        if (!array)
+            goto error;
+    }
+
+    dprintk("%s: allocated new frames array: %p\n", __func__, array);
+
+    init_frames_array(array);
+
+    return array;
+error:
+    panic("PMM: Unable to allocate new page for frame array");
+    UNREACHABLE();
+}
+
+static frames_array_t *new_frames_array_nolock(void) {
+    frames_array_t *array;
+    frame_t *frame;
+
+    frame = reserve_frame(get_first_frame(free_frames, PAGE_ORDER_4K));
+    if (!frame)
+        goto error;
+
+    if (!boot_flags.virt)
+        array = (frames_array_t *) mfn_to_virt_kern(frame->mfn);
+    else {
+        /* switch to special refilling mode to avoid deadlock with paging */
+        spin_unlock(&priority_lock);
+
+        /* only paging is allowed to use pmm with only the priority_lock */
+        array = vmap_pmm_refill_nolock(mfn_to_virt_map(frame->mfn), frame->mfn, L1_PROT);
```

Comment:
since this function is almost identical to the previous one (modulo the vmap callee, maybe define it above the previous one as a helper function (i.e. prefixed with an `_`) and a parameter specifying the callee. Then you can derive two small inline functions from it: the `new_frames_array` and the `new_frames_array_nolock`. Check the vmap_kern_4k(), it might be an inline function, so pass a pointer to the real deal.

Reply:
I can merge them again but the split was made because of this: https://github.com/KernelTestFramework/ktf/pull/345#discussion_r1899396681

Reply:
Yes, I had something else in mind. Just did not specify it back then. Sorry.